lib/services/reconciler.go

```diff
@@ -23,33 +23,43 @@ import (
 
 	"github.com/gravitational/trace"
 	"github.com/sirupsen/logrus"
-
-	"github.com/gravitational/teleport/api/types"
 )
 
 // ReconcilerConfig is the resource reconciler configuration.
-type ReconcilerConfig struct {
+type ReconcilerConfig[T any] struct {
+	GetKind   func(T) string
+	GetName   func(T) string
+	GetOrigin func(T) string
 	// Matcher is used to match resources.
-	Matcher Matcher
+	Matcher Matcher[T]
 	// GetCurrentResources returns currently registered resources.
-	GetCurrentResources func() types.ResourcesWithLabelsMap
+	GetCurrentResources func() map[string]T
```

Comment:
`[]T` would be much better than `map[string]T`, IMO, but I suppose this one already sailed.